test/server.cpp

```diff
@@ -631,157 +631,157 @@ TEST_F(ServerTest, Http404HtmlError)
+    { /* url */ "/ROOT%23%3F/catalog/invalid_endpoint?userlang=test",
       expected_page_title=="[I18N TESTING] Not Found - Try Again" &&
       expected_body==R"(
     <h1>[I18N TESTING] Content not found, but at least the server is alive</h1>
     <p>
-      [I18N TESTING] URL not found: /ROOT/catalog/invalid_endpoint
+      [I18N TESTING] URL not found: /ROOT%23%3F/catalog/invalid_endpoint
     </p>
 )"  },
 
-    { /* url */ "/ROOT/content/invalid-book/whatever",
+    { /* url */ "/ROOT%23%3F/content/invalid-book/whatever",
       expected_body==R"(
     <h1>Not Found</h1>
     <p>
-      The requested URL "/ROOT/content/invalid-book/whatever" was not found on this server.
+      The requested URL "/ROOT%23%3F/content/invalid-book/whatever" was not found on this server.
     </p>
     <p>
-      Make a full text search for <a href="/ROOT/search?pattern=whatever">whatever</a>
+      Make a full text search for <a href="/ROOT%23%3F/search?pattern=whatever">whatever</a>
     </p>
 )"  },
 
-    { /* url */ "/ROOT/content/zimfile/invalid-article",
+    { /* url */ "/ROOT%23%3F/content/zimfile/invalid-article",
       book_name=="zimfile" &&
       book_title=="Ray Charles" &&
       expected_body==R"(
     <h1>Not Found</h1>
     <p>
-      The requested URL "/ROOT/content/zimfile/invalid-article" was not found on this server.
+      The requested URL "/ROOT%23%3F/content/zimfile/invalid-article" was not found on this server.
     </p>
     <p>
-      Make a full text search for <a href="/ROOT/search?content=zimfile&pattern=invalid-article">invalid-article</a>
+      Make a full text search for <a href="/ROOT%23%3F/search?content=zimfile&pattern=invalid-article">invalid-article</a>
     </p>
 )"  },
 
-    { /* url */ R"(/ROOT/content/"><svg onload=alert(1)>)",
+    { /* url */ R"(/ROOT%23%3F/content/"><svg onload=alert(1)>)",
       expected_body==R"(
     <h1>Not Found</h1>
     <p>
-      The requested URL "/ROOT/content/&quot;&gt;&lt;svg onload=alert(1)&gt;" was not found on this server.
+      The requested URL "/ROOT%23%3F/content/%22%3E%3Csvg%20onload%3Dalert(1)%3E" was not found on this server.
     </p>
     <p>
-      Make a full text search for <a href="/ROOT/search?pattern=%22%3E%3Csvg%20onload%3Dalert(1)%3E">&quot;&gt;&lt;svg onload=alert(1)&gt;</a>
+      Make a full text search for <a href="/ROOT%23%3F/search?pattern=%22%3E%3Csvg%20onload%3Dalert(1)%3E">&quot;&gt;&lt;svg onload=alert(1)&gt;</a>
     </p>
 )"  },
 
-    { /* url */ R"(/ROOT/content/zimfile/"><svg onload=alert(1)>)",
+    { /* url */ R"(/ROOT%23%3F/content/zimfile/"><svg onload=alert(1)>)",
       book_name=="zimfile" &&
       book_title=="Ray Charles" &&
       expected_body==R"(
     <h1>Not Found</h1>
     <p>
-      The requested URL "/ROOT/content/zimfile/&quot;&gt;&lt;svg onload=alert(1)&gt;" was not found on this server.
+      The requested URL "/ROOT%23%3F/content/zimfile/%22%3E%3Csvg%20onload%3Dalert(1)%3E" was not found on this server.
```

Comment:
I would say that we need the url decoded version here.
The url entered by the user don't contain % encoding, we may need to encode it when used in url. But this is not really a url, this is text for the user.

Reply:
Fixed, though with a subtle difference from how it worked originally. I `urlDecode()` weakly so that most reserved symbols remain encoded (otherwise the URL may be misinterpreted because of the `?` or `#` symbols appearing in it). In this particular case `=` remains URI-encoded.

Reply:
I fixed it in place (rather than in a fixup commit) so that the commit message could be updated too.